lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,2460 @@
+        with torch.no_grad():
+            z_embed = self.vq_layer.get_codes_from_indices(encoding_indices)
+            z_embed = z_embed.sum(dim=0)
+        return z_embed
+
+    def get_action_from_latent(self, latent):
+        output = self.decoder(latent)
+        if self.config.action_chunk_size == 1:
+            return einops.rearrange(output, "N (T A) -> N T A", A=self.config.output_shapes["action"][0])
+        else:
+            return einops.rearrange(output, "N (T A) -> N T A", A=self.config.output_shapes["action"][0])
+
+    def preprocess(self, state):
+        if not torch.is_tensor(state):
+            state = torch.FloatTensor(state.copy())
+        if self.config.action_chunk_size == 1:
+            state = state.squeeze(-2)
+        else:
+            state = einops.rearrange(state, "N T A -> N (T A)")
+        return state
+
+    def get_code(self, state, required_recon=False):
+        state = self.preprocess(state)
+        with torch.no_grad():
+            state_rep = self.encoder(state)
+            state_rep_shape = state_rep.shape[:-1]
+            state_rep_flat = state_rep.view(state_rep.size(0), -1, state_rep.size(1))
+            state_rep_flat, vq_code, vq_loss_state = self.vq_layer(state_rep_flat)
+            state_vq = state_rep_flat.view(*state_rep_shape, -1)
+            vq_code = vq_code.view(*state_rep_shape, -1)
+            vq_loss_state = torch.sum(vq_loss_state)
+            if required_recon:
+                recon_state = self.decoder(state_vq)
+                recon_state_ae = self.decoder(state_rep)
+                if self.config.action_chunk_size == 1:
+                    return state_vq, vq_code, recon_state, recon_state_ae
+                else:
+                    return (
+                        state_vq,
+                        vq_code,
+                        torch.swapaxes(recon_state, -2, -1),
+                        torch.swapaxes(recon_state_ae, -2, -1),
+                    )
+            else:
+                return state_vq, vq_code
+
+    def vqvae_forward(self, state):
+        state = self.preprocess(state)
+        state_rep = self.encoder(state)
+        state_rep_shape = state_rep.shape[:-1]
+        state_rep_flat = state_rep.view(state_rep.size(0), -1, state_rep.size(1))
+        state_rep_flat, vq_code, vq_loss_state = self.vq_layer(state_rep_flat)
+        state_vq = state_rep_flat.view(*state_rep_shape, -1)
+        vq_code = vq_code.view(*state_rep_shape, -1)
+        vq_loss_state = torch.sum(vq_loss_state)
+
+        dec_out = self.decoder(state_vq)
+        encoder_loss = (state - dec_out).abs().mean()
+
+        rep_loss = encoder_loss + vq_loss_state * 5
+
+        metric = (
+            encoder_loss.clone().detach(),
+            vq_loss_state.clone().detach(),
+            vq_code,
+            rep_loss.item(),
+        )
+        return rep_loss, metric
+
+
+    def load_state_dict(self, *args, **kwargs):
```

Comment:
Removed `def load_state_dict`, and now `discretized` which was in `def toggle_discretized`, and `vq_layer.freeze_codebook` is now neatly included in the model state dict.

Technically, we could consolidate `discretized` and `vq_layer.freeze_codebook`, but I am not doing that; this is because vq_layer.freeze_codebook is in a very low-level subclass, and we don't want to reach that far down every time we check whether it's discretized.